### ISSUE TITLE
[Select] Question - can Enhanced mdc-select be used in SPA or is custom code needed? 

### ISSUE THREAD

**Issue by cintaccs** [material-components/material-components-web#4541]
Hi,
Thanks for providing a great component library! In general I have been able to implement most components successfully in a single page application - they clean up nicely and as expected with the exception of the "enhanced mdc-select" variant where I have the issue that the component init will re-locate the menu-surfaced list to end of the DOM body - and in my case - it will leave it there.

The problem is that when the SPA (updating specific DIV or whatever elements on the page) is changing to other "pages" the "re-located" menu-surfaced listings - still remains at the bottom of the body element. Question is whether I am using the components wrong - or that is intended behaviour. I think it causes the select components to become periodic unresponsive on focus click (each second time eventually) and the "garbage" piles up in the body. 

I think the problem is not only and issue for SPA's - but also would be for "enhanced mdc-select" components on various tab pages (you have a demo example of using multiple div elements for the tab-control) if you have enhanced select components within these tab pages - moving from one tab page to the next and back again will add "no-longer-linked" re-located "enhanced mdc select" garbage at the end of your body document.

I have previously reported this issue (#4450 ) - but it was closed - but the problem remains. 

Thanks in advance for any information or guidance you can give. I am aware that the MDC menu can be "linked" (menuSurface_.anchorElement) to a menu surface located elsewhere in the document - and on menu.open - it doesn't re-locate anything. Question is whether that same method could/should be used for Enhanced mdc select (especially in an SPA where the body of the document is only partly updated and the body "garbage" is therefore difficult to track) (solution is maybe to locate the actual option list  outside the SELECT instead of nested within)?

**Comment by moog16** [comment]
We should be able to unhoist `menu_` within enhanced select during the destroy(), which will put the menu back into the root element of the select. That way when you tear down select, this will also remove the menu. Will this work for you?

As a work around for now, you can put a unique ID on the menu, and remove it through getElementById. That should work until we address this issue. Please feel free to open a PR otherwise. :)

**Comment by cintaccs** [comment]
Thanks a lot for the feedback! I will investigate. 

I think my biggest "concern" is that the nodes are moved around in the DOM. Apart from the issue (I have...maybe alone..) I would in general consider moves in the DOM a bit "dangerous" (eventually bad for performance too) - the same logic that is applied for root vs. elements and for example the "next sibling" logic used in "mdc-notched-outline" elsewhere - could be influenced by appending contents to the end of the body (i.e. other components that are looking into the content of the body to determine state or whatever). I am not 100% sure why it is necessary (I understand it relates to scroll bar issues) but I think that should/could be solved "in-place" with css rules and additional containers under the root element maybe. OR the hoisted elements should be re-directed to a known location that can be destroyed more easily. If you have several selects - you will have n number of siblings at the end of the body - only a complete scan of elements in the body can find them - however then you need to figure out which are active on the page and which are left overs etc. etc. 
I currently consider using the menu component (also having the mdc list) and the anchorElement logic and store the list elements in another location (but still in the "virtual tab page scope") that will be destroyed automatically when replacing the parent scope node element with new content.

Thanks again for suggestions!

**Comment by cintaccs** [comment]
To unhoist menu_ during destroy() will move the select elements into the original "page scope" again - and therefore it should work - question is whether destroy() is always called - also when just overwriting the parent page root node with new content - or whether specific cleanup termination calls needs to be implemented. But with the destroy modification you suggest - that would be possible - without having to maintain or invent cross-page naming / id logic. If destroy() is always called on whatever tear down - it would be super perfect!

**Comment by kfranqueiro** [comment]
`destroy` is exposed on every MDC Web component and foundation (the component's `destroy` calls the foundation's). Beyond that, it is the responsibility of the consumer to ensure that components' `destroy` methods are called before otherwise tearing down or removing component DOM within the  host environment. (If you're using a library that wraps MDC Web, it should be calling `destroy` at the appropriate time.)

**Comment by abhiomkar** [comment]
@cintaccs Thanks for detailed explanation! 

As Ken mentioned above `destroy()` method has to be called by consumer on end-of-life of the component.

Regarding your original issue, do you think exposing `hoistMenuToBody` boolean API to public would solve your problem?

Select component need to provide hoisted (when rendered inside dialog) & unhoisted option (to avoid dropdown being detached from anchor on scroll).

**Comment by cintaccs** [comment]
@abhiomkar Thanks for the feedback!

I am not 100% sure whether it will help. hoistMenuToBody in itself takes no parameters as I see it - and according to the description - is actually doing exactly what causes the issue (remove from DOM and append to body) - but maybe if used with setMenuSurfaceAnchorElement(element: Element)... from my point of view it would be nice to have the ability to control which DOM node to append to (i.e. to overwrite the "append to body" behaviour).

I think it is fair to assume that elements are destroy'ed - however (to my knowledge) - all other elements are "auto" destroyed and "auto" garbage collected by simply removing any parent DOM nodes that the components are created below - and especially the DOM remove/append behaviour is breaking that "rule" when only updating parts of a page (as in a SPA). Whatever solution that require "manual" recording and destruction of each single component instance  will be an additional complexity overhead.

Maybe I am wrong - but I am using pretty much all of the MDC components in my app - and only the enhanced select list seems to require this manual destroy. To me it reminds of the "old" register and deregister of eventhandlers that used to be best practice - but seems automatically handled well by browsers nowadays.

From my point of view the select component in MDC have a few issues however:
- you have to fill the select list at component creation time (i.e. you have to know and fill the enteries before you know whether the user ever click or activate the select list).
- you can't "search" the list (i.e. there is no UI interface for entering search terms - actually the root element of the select list could be used for capturing input for search...)
- you can't rely on the automatic clean up (because of the DOM append to body - outside of the content that it was originally created within).
- if you have the same list basis (the same selection list) required on multiple UI components (like when the table component will be created for the MDC library and a select is behind any of the columns of the table (https://material.io/design/components/data-tables.html#behavior and especially "An expanded inline menu") is needed - then the enhanced select can't be used without repeating the content many times in the DOM.
- If you have many items in the enhanced select list (or even a non-ending list of items) it can't logically be created at the time of the element init - it must be resolved at user interaction.
- The nested - waterfall - type selection list might be very complex at creation time....

I might have misunderstood something about how to use the enhanced select - or maybe the "menu" component is actually what should be used instead ...  

I understand that the component library is complex and I don't know all the details - so maybe I am wrong - but from my point of view the implemented solution to "avoid dropdown being detached from anchor on scroll" might be the cause of the problem...

**Comment by abhiomkar** [comment]
These are definitely valid and reasonable feature requests for select component, I would suggest opening a new issue for each feature requests so it can be tracked separately.

Coming back to your original issue, as per current implementation `hoistMenuToBody` is a boolean flag, when true it moves the dropdown DOM from current location to body element (this is default behaviour of select as of now), when false it does not change the DOM location (I think this is what you wanted?).

My proposal is to expose this API to public on select component so that consumer can override it based on the need. I also agree that select component itself shouldn't move the DOM location to avoid any side-effects.

**Comment by cintaccs** [comment]
I agree - it sounds like hoistMenuToBody boolean flag is the one.

Should that be possible right now? 

select.menu_.menuSurface_.foundation_.isHoistedElement_ = false;

doesn't seems to have any impact on the behaviour - apart from the select list no longer appears :-)

**Comment by abhiomkar** [comment]
```ts
this.menu_.hoistMenuToBody();
```

The above method shouldn't be called on select instantiation ([code ref](https://github.com/material-components/material-components-web/blob/6d9c0cc43b7ab5f820bec2c95410b31a396d7587/packages/mdc-select/component.ts#L366)), it should be called on demand instead.

**mdc-select/component.ts**

```ts
hoistMenuToBody() {
  if (this.menu_) {
    this.menu_.hoistMenuToBody();
  } else {
    throw Error('MDCSelect: hoistMenuToBody is applicable to enhanced select variant only.');
  }
}
```

Since this issue is already in our backlog it'll be picked up when we've bandwidth to work on this. Feel free to open a pull request. :)

**Comment by cintaccs** [comment]
@abhiomkar hi, I have just updated to 2.0.0 and was just checking where you are on this issue. I can see that you have closed #879 with reference to this (#4541) issue. If you open a menu (not the enhanced select) then resize the screen to force scroll bars - then you can scroll and the "menu" will detach from it's original target (which IMO it shouldn't). I think #879 might be pointing to that problem - which I think is unrelated to the "hoistMenuToBody" of enhanced selects in #4541. I am wondering whether it however all relates to the same problem - that scrolling is not handled correctly (which leads to the (IMO wrong solution) of relocating the select list DOM node - and still seems to be an issue for the menu-surface of the menu component also. As far as I can tell - the MDC solution for scrim area around the dialogue component is correct in terms of ignoring window scrolling outside the dialogue window. Could the same logic as used for the scrim area of dialoques not be used for menu-surface and enhanced select lists? I am aware that the scrim for dialoques are dark opacity - but it could be set to something almost invisible like
background: rgba(255,255,255,0.0005);
In other words - could it be a better solution to use the scrim logic of dialoques for menu and enhanced select lists as you have currently implemented for the dialoque component - but with an almost invisible cover background as mentioned above?
